ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision/Common.hs

```diff
@@ -0,0 +1,677 @@
+chain   C         A         B         A         D         B         A
+
+This is the form in which we are informed about candidate chains from the
+consensus layer, the combination of a chain and the peer it is from. This
+makes sense, since these things change independently.
+
+We will process the chains in this form, keeping the peer/chain combination all
+the way through. Although there could in principle be some opportunistic saving
+by sharing when multiple peers provide the same chain, taking advantage of this
+adds complexity and does nothing to improve our worst case costs.
+
+We are only interested in candidate chains that are strictly longer than our
+current chain. So our first task is to filter down to this set.
+-}
+
+-- | Keep only those candidate chains that are preferred over the current
+-- chain. Typically, this means that their length is longer than the length of
+-- the current chain.
+--
+filterPlausibleCandidates
+  :: (AnchoredFragment block -> AnchoredFragment header -> Bool)
+  -> AnchoredFragment block  -- ^ The current chain
+  -> [(AnchoredFragment header, peerinfo)]
+  -> [(FetchDecision (AnchoredFragment header), peerinfo)]
+filterPlausibleCandidates plausibleCandidateChain currentChain chains =
+    [ (chain', peer)
+    | (chain,  peer) <- chains
+    , let chain' = do
+            guard (plausibleCandidateChain currentChain chain)
+              ?! FetchDeclineChainNotPlausible
+            return chain
+    ]
+
+{-
+We define the /fetch range/ as the suffix of the fork range that has not yet
+had its blocks downloaded and block content checked against the headers.
+
+    ┆   ┆
+    ├───┤
+    │   │
+    ├───┤               ┌───┐
+    │   │    already    │   │
+    ├───┤    fetched    ├───┤
+    │   │    blocks     │   │
+    ├───┤               ├───┤
+    │   │               │░◉░│  ◄  fetch range
+ ───┴───┴─────┬───┬─────┼───┼───
+              │░◉░│ ◄   │░░░│
+              └───┘     ├───┤
+                        │░◉░│  ◄
+                        └───┘
+
+In earlier versions of this scheme we maintained and relied on the invariant
+that the ranges of fetched blocks are backwards closed. This meant we never had
+discontinuous ranges of fetched or not-yet-fetched blocks. This invariant does
+simplify things somewhat by keeping the ranges continuous however it precludes
+fetching ranges of blocks from different peers in parallel.
+
+We do not maintain any such invariant and so we have to deal with there being
+gaps in the ranges we have already fetched or are yet to fetch. To keep the
+tracking simple we do not track the ranges themselves, rather we track the set
+of individual blocks without their relationship to each other.
+
+-}
+
+-- | Find the fragments of the chain suffix that we still need to fetch, these
+-- are the fragments covering blocks that have not yet been fetched and are
+-- not currently in the process of being fetched from this peer.
```

Comment:
I'm failing to see see how the not-inflight blocks are determined in this function.

Reply:
This comment comes from before; cf https://github.com/IntersectMBO/ouroboros-network/blob/49150d2679b0cca15ca7238b75b696abad6a53d9/ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs#L571-L576 on `main`. I think it is supposed to apply to both functions `filterNotAlreadyFetched` and `filterNotAlreadyInFlightWithPeer`. I am not necessarily very happy with this either and now is our chance to change it.

Reply:
Fixed in d4a453db395bda67b848b8799a8a5fff328cc09a.